_posts/2017-10-30-user-secrets.md

```diff
@@ -0,0 +1,89 @@
+
+
+##### Create and Reference a User Secret
+
+The next step is to create a user secret containing the sensitive data and reference it in the project.
+<br/>
+In VS you can right click on the project file and you should see a menu item called "Manage User Secrets". However, it is more interesting to do it manually at least once to understand how it works.
+
+A user secret is just a JSON file. It must be stored under `~/.microsoft/usersecrets/<userSecretsId>/secrets.json` (`%APPDATA%\microsoft\UserSecrets\<userSecretsId>\secrets.json` on Windows) where `<userSecretsId>` can be anything.
+<br/>
+In my case my user secret looks like:
+```
+// ~/.microsoft/usersecrets/serviceConfigSecretId/secrets.json
+{
+  "SecretServiceConfig": {
+    "address": "https://somewhere-secret.com",
+    "key": "super-secret-key"
+  }
+}
+```
+
+Then, the user secret must be referenced in the project file:
+```xml
+<PropertyGroup>
+    <TargetFramework>netcoreapp2.0</TargetFramework>
+    <UserSecretsId>serviceConfigSecretId</UserSecretsId>
+</PropertyGroup>
+```
+
+With that in place, the user secret can now be used in code.
+
+
+##### Add the User Secret to the Configuration at run time
+
+Ultimately, when my application will be deployed, the sensitive data will be injected via the configuration at start time. So what I want is to be able to add the User Secrets to the configuration at start time but only in development mode. The way (or at least one way) to do it to add the user secret to the C# Configuration object when it is built:
+```
+public class Startup
+{
+    public Startup(IHostingEnvironment env)
+    {
+        var builder = new ConfigurationBuilder()
+            .SetBasePath(env.ContentRootPath)
+            .AddJsonFile("appsettings.json");
+
+        if (env.IsDevelopment())
+        {
+            builder.AddUserSecrets<Startup>();
+        }
+
+        Configuration = builder.Build();
+    }
+```
+An interesting thing to note here is that one can mix configuration sources. In my case, my application configuration comes from two sources: one is the appsettings.json file stored with my code; the other is the data stored in the user secrets. This is super useful!
+
+Once the configuration object is built, it doesn't matter where the data came from, you can access it as usual by calling `Configuration.GetSection("SecretServiceConfig")`. This will return a ConfigurationSection object with all the data.
+
+
+##### Using Option to access Configuration
+
+The Configuration object can be used directly but a more elegant way to access configuration is using what Microsoft calls the option pattern. It basically helps match configuration section to C# classes. I won't describe how I used it in this post as it is out of scope but it is worth knowing about it, you can check out this [article](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/configuration?tabs=basicconfiguration) about Configuration, they mention it.
```

Comment:
Are there multiple configuration sections? If so it should be 'match configuration section_s_' , if it's singular then it should be '_the_ configuration section_'

Reply:
yes, there are.